test/user_test.rb

```diff
@@ -0,0 +1,46 @@
+require 'minitest/autorun'
+require 'minitest/pride'
+require './lib/joke'
+require './lib/user'
+
+class UserTest < Minitest::Test
+  def test_user_exists
+    sal = User.new("Sal")
+    assert_instance_of User, sal
+  end
+
+  def test_can_learn_joke
+    sal = User.new("Sal")
+    joke_1 = Joke.new(1, "Why did the strawberry cross the road?", "Because his mother was in a jam.")
+    joke_2 = Joke.new(2, "How do you keep a lion from charging?", "Take away its credit cards.")
+    sal.learn(joke_1)
+    sal.learn(joke_2)
+    assert_equal sal.jokes.length, 2
+  end
+
+  def test_can_tell_joke
+    sal = User.new("Sal")
+    ali = User.new("Ali")
+    joke_1 = Joke.new(1, "Why did the strawberry cross the road?", "Because his mother was in a jam.")
+    joke_2 = Joke.new(2, "How do you keep a lion from charging?", "Take away its credit cards.")
+    sal.learn(joke_1)
+    sal.learn(joke_2)
+    sal.tell(ali, joke_1)
+    sal.tell(ali, joke_2)
+
+    assert_equal sal.jokes, ali.jokes
+  end
+
+  def test_categorized_by_id
+    sal = User.new("Sal")
+    ali = User.new("Ali")
+    joke_1 = Joke.new(1, "Why did the strawberry cross the road?", "Because his mother was in a jam.")
+    joke_2 = Joke.new(2, "How do you keep a lion from charging?", "Take away its credit cards.")
+    sal.learn(joke_1)
+    sal.learn(joke_2)
+    sal.tell(ali, joke_1)
+    sal.tell(ali, joke_2)
+    assert_equal 1, ali.joke_by_id(1).first.id
```

Comment:
this is the hack i was telling you about--- you're calling `.first` on the method return value here to get the actual joke, which we shouldn't have to do, according to the interaction pattern. 

A true test, according to the expected behavior from the interaction pattern, would be: 
`assert_equal joke_1, ali.joke_by_id(1)`
which would have failed, the way your method is currently written (which is what we want! we want tests that enforce expected behavior)